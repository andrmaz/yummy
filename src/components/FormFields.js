import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { apiCall } from '../api/apiCall';
import SelectRecipes from './SelectRecipes';
import SelectDiet from './SelectDiet';
import SelectHealth from './SelectHealth';
import { 
    Center,
    InputGroup, 
    InputRightElement, 
    Button, 
    Flex, 
    FormControl, 
    useColorMode,
} from '@chakra-ui/react';
// import dotenv configuration file
import '../config';
import SearchInput from './SearchInput';


const app_id = process.env.REACT_APP_ID;
const app_key = process.env.REACT_APP_KEY;


const FormFields = ({ onSearch }) => {

    // To manage color mode chakra exposes the useColorMode
    const { colorMode } = useColorMode();

    const { control, handleSubmit, watch, errors } = useForm({
		mode: 'onChange',
        reValidateMode: 'onChange',
        defaultValues: {
            input: '',
            to: 4,
            diet: 'balanced',
            health: 'tree-nut-free',
        }
    });

    // Check on props change
    const { input, to, diet, health }  = watch();

    
    const onSubmit = async () => {
        try {
            const options = {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'Content-Encoding': 'gzip',
                    'Cache-Control': 'no-cache'
                }
            };
            const uri = `https://api.edamam.com/search?q=${input}&app_id=${app_id}&app_key=${app_key}&from=0&to=${to}&diet=${diet}&health=${health}`;
            const data = await apiCall(uri, options);
            // if no data match the search request, return an object with count 0
            onSearch(data);
        } catch (error) {
            throw new Error(error.message);
        }
    };

    return (
        <Center h="20vh" color={colorMode === 'light' ? 'black' : 'white'} minW='300px' maxW='500px' m='auto'>
            <FormControl onSubmit={handleSubmit(onSubmit)} isRequired='true' >
                <InputGroup size='md'>
                    <SearchInput control={control} errors={errors} />
                    <InputRightElement width='4.5rem'>
                        <Button 
                            id='button'
                            h='1.75rem' 
                            size='sm'
                            variant='solid'
                            colorScheme='blue'
                            isDisabled={input.length < 2 ? true : false}
                            onClick={onSubmit}
                        >
                            Search
                        </Button>
                    </InputRightElement>
                </InputGroup>
                <Flex justify='space-around'>
                    <SelectRecipes control={control} />
                    <SelectDiet control={control} />
                    <SelectHealth control={control} />
                </Flex>
            </FormControl>
        </Center>
    );
};

FormFields.propTypes = {
    onSearch: PropTypes.func,
};

export default FormFields;
