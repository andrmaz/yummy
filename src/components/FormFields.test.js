import React from 'react';
import { shallow } from 'enzyme';
import FormFields from './FormFields';

let wrapper;
const mockSearch = jest.fn();
const mockValues = {
    diet: 'balanced',
    health: 'tree-nut-free',
    input: 'chicken',
    to: 4,
};

describe('FormFields Component', () => {
    beforeEach(() => {
        wrapper = shallow(<FormFields onSearch={mockSearch} />);
        expect.hasAssertions();
    });
    it('should render Component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should not click button on empty input', () => {
        const element = wrapper.find('[id="button"]');
        element.simulate('click');
        expect(mockSearch).not.toBeCalled();
    });
});
