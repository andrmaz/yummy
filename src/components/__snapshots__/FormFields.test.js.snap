// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFields Component should render Component correctly 1`] = `
<Center
  color="white"
  h="20vh"
  m="auto"
  maxW="500px"
  minW="300px"
>
  <FormControl
    isRequired="true"
    onSubmit={[Function]}
  >
    <InputGroup
      size="md"
    >
      <SearchInput
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
        errors={Object {}}
      />
      <InputRightElement
        width="4.5rem"
      >
        <Button
          colorScheme="blue"
          h="1.75rem"
          id="button"
          isDisabled={true}
          onClick={[Function]}
          size="sm"
          variant="solid"
        >
          Search
        </Button>
      </InputRightElement>
    </InputGroup>
    <Flex
      justify="space-around"
    >
      <SelectRecipes
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
      />
      <SelectDiet
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
      />
      <SelectHealth
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
      />
    </Flex>
  </FormControl>
</Center>
`;
